### ISSUE TITLE
Stream resumption crash some times and some times works 

### ISSUE THREAD

**Issue by jaspreet-android** [esl/MongooseIM#3076]
MongooseIM version: 3.7.1
Installed from: source
Erlang/OTP version: 22

Describe the issue.

Some time user session is terminated on stream resumption. 

Following are logs for the same in ejabberd.log 

```
2021-04-08 02:46:08.548 [info] <0.1723.0>@mongoose_tcp_listener:accept_loop:91 event=accepted_tcp_connection, socket=#Port<0.8998>, details=#{dest_address => {10,10,7,30},dest_port => 5222,proxy => false,src_address => {10,10,7,33},src_port => 60516}
2021-04-08 02:46:30.921 [info] <0.16190.1>@ejabberd_c2s:handle_sasl_success:3253 ({socket_state,ejabberd_tls,{ejabberd_tls_socket,fast_tls,#Port<0.8998>,{tlssock,#Port<0.8998>,#Ref<0.1636462823.3076653058.12979>},[{ciphers,[84,76,83,118,49,46,50,58,84,76,83,118,49,46,51]},{protocol_options,[110,111,95,115,115,108,118,50,124,110,111,95,115,115,108,118,51,124,110,111,95,116,108,115,118,49,124,110,111,95,116,108,115,118,49,95,49]},verify_none,{certfile,[47,109,111,110,103,111,111,115,101,105,109,47,97,108,108,46,112,101,109]}],false},<0.16189.1>,#{dest_address => {10,10,7,30},dest_port => 5222,proxy => false,src_address => {10,10,7,33},src_port => 60516}}) Accepted authentication for cde5ab58-eb0f-4e35-8828-26409a873039 by ejabberd_auth_rdbms
2021-04-08 02:46:31.502 [info] <0.16190.1>@ejabberd_c2s:do_resume_session:3120 event=resumption_error reason=no_previous_session_for_smid smid=<<"sGaXrOzeuNf8UoqAngQCd0eONyyJ">>
2021-04-08 02:46:32.301 [info] <0.16190.1>@ejabberd_c2s:do_open_session:757 ({socket_state,ejabberd_tls,{ejabberd_tls_socket,fast_tls,#Port<0.8998>,{tlssock,#Port<0.8998>,#Ref<0.1636462823.3076653058.12979>},[{ciphers,[84,76,83,118,49,46,50,58,84,76,83,118,49,46,51]},{protocol_options,[110,111,95,115,115,108,118,50,124,110,111,95,115,115,108,118,51,124,110,111,95,116,108,115,118,49,124,110,111,95,116,108,115,118,49,95,49]},verify_none,{certfile,[47,109,111,110,103,111,111,115,101,105,109,47,97,108,108,46,112,101,109]}],false},<0.16189.1>,#{dest_address => {10,10,7,30},dest_port => 5222,proxy => false,src_address => {10,10,7,33},src_port => 60516}}) Opened session for cde5ab58-eb0f-4e35-8828-26409a873039@localhost/resource
2021-04-08 02:46:33.352 [info] <0.16190.1>@mod_carboncopy:iq_handler:100 carbons enabled for user cde5ab58-eb0f-4e35-8828-26409a873039@localhost/resource
2021-04-08 02:53:40.937 [info] <0.16190.1>@ejabberd_c2s:terminate:1540 ({socket_state,ejabberd_tls,{ejabberd_tls_socket,fast_tls,#Port<0.8998>,{tlssock,#Port<0.8998>,#Ref<0.1636462823.3076653058.12979>},[{ciphers,[84,76,83,118,49,46,50,58,84,76,83,118,49,46,51]},{protocol_options,[110,111,95,115,115,108,118,50,124,110,111,95,115,115,108,118,51,124,110,111,95,116,108,115,118,49,124,110,111,95,116,108,115,118,49,95,49]},verify_none,{certfile,[47,109,111,110,103,111,111,115,101,105,109,47,97,108,108,46,112,101,109]}],false},<0.16189.1>,#{dest_address => {10,10,7,30},dest_port => 5222,proxy => false,src_address => {10,10,7,33},src_port => 60516}}) Stream <<"ngE/3UN2NK1intmLlmoAyo+iq8kK">> resumed for cde5ab58-eb0f-4e35-8828-26409a873039@localhost/resource
2021-04-08 02:53:40.938 [error] <0.16190.1> gen_fsm <0.16190.1> in state session_established terminated with reason: {handover_session,{<18634.1407.2>,#Ref<18634.3960687319.1540358146.64388>}}
2021-04-08 02:53:40.938 [error] <0.16190.1> CRASH REPORT Process <0.16190.1> with 0 neighbours exited with reason: {handover_session,{<18634.1407.2>,#Ref<18634.3960687319.1540358146.64388>}} in p1_fsm_old:terminate/8 line 733
2021-04-08 02:53:40.938 [error] <0.620.0> Supervisor ejabberd_c2s_sup had child undefined started with {ejabberd_c2s,start_link,undefined} at <0.16190.1> exit with reason {handover_session,{<18634.1407.2>,#Ref<18634.3960687319.1540358146.64388>}} in context child_terminated
```

Following are logs for the same in crash.log 

```
2021-04-08 02:53:40 =ERROR REPORT====
** State machine <0.16190.1> terminating
** Last message in was {'$gen_sync_event',{<18634.1407.2>,#Ref<18634.3960687319.1540358146.64388>},resume}
** When State == session_established
**      Data  == {state,{socket_state,ejabberd_tls,{ejabberd_tls_socket,fast_tls,#Port<0.8998>,{tlssock,#Port<0.8998>,#Ref<0.1636462823.3076653058.12979>},[{ciphers,"TLSv1.2:TLSv1.3"},{protocol_options,"no_sslv2|no_sslv3|no_tlsv1|no_tlsv1_1"},verify_none,{certfile,"/mongooseim/all.pem"}],false},<0.16189.1>,#{dest_address => {10,10,7,30},dest_port => 5222,proxy => false,src_address => {10,10,7,33},src_port => 60516}},ejabberd_socket,#Ref<0.1636462823.3078881281.227457>,false,<<"3905E16D50D071C3">>,{sasl_state,<<"jabber">>,<<"localhost">>,<<>>,cyrsasl_scram_sha1,{state,4,<<164,113,186,212,190,167,174,36,112,38,57,72,42,0,184,64,213,81,1,255>>,<<71,29,208,202,168,228,216,132,31,217,68,193,198,186,248,193,140,166,205,109>>,<<"cde5ab58-eb0f-4e35-8828-26409a873039">>,{mongoose_credentials,<<"localhost">>,[],[]},<<"n=cde5ab58-eb0f-4e35-8828-26409a873039,r=K6Z2VX4T8OQ4Z_DjGDPYCCVZc_lUd6zlascsXwHDwPsilxYvhc5aORsYUFpJLx9yCnIkh7qvypkK52twMdIfO1tGdAJBw7PExX180mATNswwZAZZzfOQ8sXb2op3GE_hrku25Q,r=K6Z2VX4T8OQ4Z_DjGDPYCCVZc_lUd6zlascsXwHDwPsilxYvhc5aORsYUFpJLx9yCnIkh7qvypkK52twMdIfO1tGdAJBw7PExX180mATNswwZAZZzfOQ8sXb2op3GE_hrku25QePBudClloIxq2knvkOXLAQ==,s=4X9qiiJLLuctlZU+0mJhJg==,i=10000">>,<<"K6Z2VX4T8OQ4Z_DjGDPYCCVZc_lUd6zlascsXwHDwPsilxYvhc5aORsYUFpJLx9yCnIkh7qvypkK52twMdIfO1tGdAJBw7PExX180mATNswwZAZZzfOQ8sXb2op3GE_hrku25Q">>,<<"ePBudClloIxq2knvkOXLAQ==">>,ejabberd_auth_rdbms,sha,none,false,<<>>},{mongoose_credentials,<<"localhost">>,[],[]}},c2s,c2s_shaper,{true,10000},true,true,true,[verify_none,{certfile,"/mongooseim/all.pem"}],verify_none,true,{jid,<<"cde5ab58-eb0f-4e35-8828-26409a873039">>,<<"localhost">>,<<"resource">>,<<"cde5ab58-eb0f-4e35-8828-26409a873039">>,<<"localhost">>,<<"resource">>},<<"cde5ab58-eb0f-4e35-8828-26409a873039">>,<<"localhost">>,<<"resource">>,{{1617,835592,305089},<0.16190.1>},{pres_t,52},{pres_f,52},{pres_a,52},{pres_i,0},[],{xmlel,<<"presence">>,[{<<"xml:lang">>,<<"en">>}],[{xmlel,<<"show">>,[],[{xmlcdata,<<"away">>}]},{xmlel,<<"priority">>,[],[{xmlcdata,<<"55">>}]},{xmlel,<<"idle">>,[{<<"since">>,<<"2021-04-07T13:18:03.448Z">>},{<<"xmlns">>,<<"urn:xmpp:idle:1">>}],[]}]},undefined,1617835593,false,{userlist,none,[],false},c2s_tls,ejabberd_auth_rdbms,{{10,10,7,33},60516},[{caps_resources,{0,nil}}],<<"en">>,true,18,<<"ngE/3UN2NK1intmLlmoAyo+iq8kK">>,18,[],0,100,1,10,undefined,undefined,undefined,active,[],0,[],#{pep_message => empty_state}}
** Reason for termination =
** {handover_session,{<18634.1407.2>,#Ref<18634.3960687319.1540358146.64388>}}
2021-04-08 02:53:40 =CRASH REPORT====
  crasher:
    initial call: ejabberd_c2s:init/1
    pid: <0.16190.1>
    registered_name: []
    exception exit: {{handover_session,{<18634.1407.2>,#Ref<18634.3960687319.1540358146.64388>}},[{p1_fsm_old,terminate,8,[{file,"/mongooseim/src/p1_fsm_old.erl"},{line,733}]},{proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,259}]}]}
    ancestors: [ejabberd_c2s_sup,ejabberd_sup,<0.439.0>]
    message_queue_len: 0
    messages: []
    links: [<0.620.0>]
    dictionary: [{'$internal_queue_len',0},{rand_seed,{#{bits => 58,jump => #Fun<rand.21.53802439>,next => #Fun<rand.18.53802439>,type => exsss,uniform => #Fun<rand.19.53802439>,uniform_n => #Fun<rand.20.53802439>},[160183361714722446|283198746862222349]}}]
    trap_exit: false
    status: running
    heap_size: 6772
    stack_size: 27
    reductions: 669266
  neighbours:
2021-04-08 02:53:40 =SUPERVISOR REPORT====
     Supervisor: {local,ejabberd_c2s_sup}
     Context:    child_terminated
     Reason:     {handover_session,{<18634.1407.2>,#Ref<18634.3960687319.1540358146.64388>}}
     Offender:   [{pid,<0.16190.1>},{id,undefined},{mfargs,{ejabberd_c2s,start_link,undefined}},{restart_type,temporary},{shutdown,brutal_kill},{child_type,worker}]

```

**Comment by imcyee** [comment]
Hi~ this is actually a closed issue. Check https://github.com/esl/MongooseIM/issues/2933

**Comment by jaspreet-android** [comment]
Thanks @imcyee I will apply fix and test it.

**Comment by jaspreet-android** [comment]
@imcyee  I am still testing it but by looking at the code it will still disconnect the user session but will prevent the crash.

**Comment by jaspreet-android** [comment]
@imcyee after applying the changes I can't find the crash yet, but I still don't understand why it kill user session during that time.

**Comment by imcyee** [comment]
@jaspreet-android My 2 cents is: it is killing the previous session, and start a new one for you. `handover_session`

https://github.com/esl/MongooseIM/blob/fad3ff71a83499b21a282fe9e27ae00b88fd6fe5/src/ejabberd_c2s.erl#L3183-L3192

**Comment by leszke** [comment]
@jaspreet-android I looked at the logs you provided and your case indeed seems to be similar to already closed #2933, like @imcyee mentioned.
You had a session with `<0.16190.1>` pid which has been alive for at least 7 min and was replaced by another one with `<18634.1407.2>` pid. It looks like this error is the same one as the one already fixed in #2933. Could you confirm the issue is not present in current MIM version?

**Comment by jaspreet-android** [comment]
Hi @leszke I am not getting crash yes. 

I think the problem is we have fixed jid lresource in the app as we don't want user to be logged in in 2 places. So we provide same resource always on xmpp bind and that is why session is replaced by another session.
